### PR TITLE
WIP: IDL introspection: yarpc meta

### PR DESCRIPTION
**Pull request by bombela**
A new yarpc meta idl handler is added to expose the generic IDL introspection.

This handler accepts a json request and returns json as well.

Without any argument, a flatmap of the IDL tree is returned.

With the `entrypoint` argument set, a flapmap of the `entrypoint` and its
recursive included IDLs. Finally, the `select` argument simply return every IDL
file matching the selection.

### CODE REVIEW

**Comment by bombela** [comment]
Examples:

```sh
yab -p http://localhost:24042 -e json keyvalue yarpc::idls | jq '.body'
{
  "idls": [
    {
      "filePath": "many/services.thrift",
      "includes": [
        "many/exceptions.thrift",
        "many/unions.thrift"
      ],
      "rawContent": "include \"./unions.thrift\"\ninclude \"./exceptions.thrift\"\n\ntypedef string Key\n\nexception InternalError {\n    1: optional string message\n}\n\nservice KeyValue {\n    // void and no exceptions\n    void setValue(1: Key key, 2: unions.ArbitraryValue value)\n\n    void setValueV2(\n        1: required Key key,\n        2: required unions.ArbitraryValue value,\n    )\n\n    // Return with exceptions\n    unions.ArbitraryValue getValue(1: Key key)\n        throws (1: exceptions.DoesNotExistException doesNotExist)\n\n    // void with exceptions\n    void deleteValue(1: Key key)\n        throws (\n            1: exceptions.DoesNotExistException doesNotExist,\n            2: InternalError internalError\n        )\n\n    list<unions.ArbitraryValue> getManyValues(\n        1: list<Key> range  // < reserved keyword as an argument\n    ) throws (\n        1: exceptions.DoesNotExistException doesNotExist,\n    )\n\n    i64 size()  // < primitve return value\n}\n\nservice Cache {\n    oneway void clear()\n    oneway void clearAfter(1: i64 durationMS)\n}\n\nstruct ConflictingNames_SetValue_Args {\n    1: required string key\n    2: required binary value\n}\n\nservice ConflictingNames {\n    void setValue(1: ConflictingNames_SetValue_Args request)\n}\n\nservice non_standard_service_name {\n    void non_standard_function_name()\n}\n",
      "sha1": "6c35f978178c675609423ce96f1536abbbf26aef"
    },
    {
      "filePath": "many/exceptions.thrift",
      "rawContent": "exception EmptyException {}\n\nexception DoesNotExistException {\n    1: required string key\n    2: optional string Error (go.name=\"Error2\")\n}\n",
      "sha1": "7515e1a7bcd9ef547ad37a0eaaeaf89df21897fa"
    },
    {
      "filePath": "many/unions.thrift",
      "includes": [
        "many/td/typedefs.thrift"
      ],
      "rawContent": "include \"./td/typedefs.thrift\"\n\nunion EmptyUnion {}\n\nunion Document {\n    1: typedefs.PDF pdf\n    2: string plainText\n}\n\nunion ArbitraryValue {\n    1: bool boolValue\n    2: i64 int64Value\n    3: string stringValue\n    4: list<ArbitraryValue> listValue\n    5: map<string, ArbitraryValue> mapValue\n}\n",
      "sha1": "873601b7a2c68a3139e7f31e054afcc403a204a4"
    },
    {
      "filePath": "many/td/typedefs.thrift",
      "includes": [
        "many/td/foo/bar/enums.thrift",
        "many/td/structs.thrift"
      ],
      "rawContent": "include \"./structs.thrift\"\ninclude \"./foo/bar/enums.thrift\"\n\ntypedef i64 Timestamp  // alias of primitive\ntypedef string State\n\ntypedef i128 UUID  // alias of struct\n\ntypedef list<Event> EventGroup  // alias fo collection\n\nstruct i128 {\n    1: required i64 high\n    2: required i64 low\n}\n\nstruct Event {\n    1: required UUID uuid  // required typedef\n    2: optional Timestamp time  // optional typedef\n}\n\nstruct Transition {\n    1: required State fromState\n    2: required State toState\n    3: optional EventGroup events\n}\n\ntypedef binary PDF  // alias of []byte\n\ntypedef set<structs.Frame> FrameGroup\n\ntypedef map<structs.Point, structs.Point> PointMap\n\ntypedef set<binary> BinarySet\n\ntypedef map<structs.Edge, structs.Edge> EdgeMap\n\ntypedef enums.EnumWithValues MyEnum\n",
      "sha1": "7e9ec84cec712c09f6b8255d722960fe96fa35a4"
    },
    {
      "filePath": "many/td/foo/bar/enums.thrift",
      "rawContent": "enum EmptyEnum {}\n\nenum EnumDefault {\n    Foo, Bar, Baz\n}\n\nenum EnumWithValues {\n    X = 123,\n    Y = 456,\n    Z = 789,\n}\n\nenum EnumWithDuplicateValues {\n    P, // 0\n    Q = -1,\n    R, // 0\n}\n\n// enum with item names conflicting with those of another enum\nenum EnumWithDuplicateName {\n    A, B, C, P, Q, R, X, Y, Z\n}\n\n// Enum treated as optional inside a struct\nstruct StructWithOptionalEnum {\n    1: optional EnumDefault e\n}\n\nenum RecordType {\n  NAME,\n  HOME_ADDRESS,\n  WORK_ADDRESS\n}\n\nenum lowerCaseEnum {\n    containing, lower_case, items\n}\n",
      "sha1": "cfb1ae035ecf1225722ed73500dbaad7e30bb557"
    },
    {
      "filePath": "many/td/structs.thrift",
      "includes": [
        "many/td/foo/bar/enums.thrift"
      ],
      "rawContent": "include \"./foo/bar/enums.thrift\"\n\nstruct EmptyStruct {}\n\n//////////////////////////////////////////////////////////////////////////////\n// Structs with primitives\n\nstruct PrimitiveRequiredStruct {\n    1: required bool boolField\n    2: required byte byteField\n    3: required i16 int16Field\n    4: required i32 int32Field\n    5: required i64 int64Field\n    6: required double doubleField\n    7: required string stringField\n    8: required binary binaryField\n}\n\nstruct PrimitiveOptionalStruct {\n    1: optional bool boolField\n    2: optional byte byteField\n    3: optional i16 int16Field\n    4: optional i32 int32Field\n    5: optional i64 int64Field\n    6: optional double doubleField\n    7: optional string stringField\n    8: optional binary binaryField\n}\n\n//////////////////////////////////////////////////////////////////////////////\n// Nested structs (Required)\n\nstruct Point {\n    1: required double x\n    2: required double y\n}\n\nstruct Size {\n    1: required double width\n    2: required double height\n}\n\nstruct Frame {\n    1: required Point topLeft\n    2: required Size size\n}\n\nstruct Edge {\n    1: required Point startPoint\n    2: required Point endPoint\n}\n\nstruct Graph {\n    1: required list<Edge> edges\n}\n\n//////////////////////////////////////////////////////////////////////////////\n// Nested structs (Optional)\n\nstruct ContactInfo {\n    1: required string emailAddress\n}\n\nstruct User {\n    1: required string name\n    2: optional ContactInfo contact\n}\n\n//////////////////////////////////////////////////////////////////////////////\n// self-referential struct\n\ntypedef Node List\n\nstruct Node {\n    1: required i32 value\n    2: optional List tail\n}\n\n\n//////////////////////////////////////////////////////////////////////////////\n// Default values\n\nstruct DefaultsStruct {\n    1: required i32 requiredPrimitive = 100\n    2: optional i32 optionalPrimitive = 200\n\n    3: required enums.EnumDefault requiredEnum = enums.EnumDefault.Bar\n    4: optional enums.EnumDefault optionalEnum = 2\n\n    5: required list<string> requiredList = [\"hello\", \"world\"]\n    6: optional list<double> optionalList = [1, 2.0, 3]\n\n    7: required Frame requiredStruct = {\n        \"topLeft\": {\"x\": 1, \"y\": 2},\n        \"size\": {\"width\": 100, \"height\": 200},\n    }\n    8: optional Edge optionalStruct = {\n        \"startPoint\": {\"x\": 1, \"y\": 2},\n        \"endPoint\":   {\"x\": 3, \"y\": 4},\n    }\n}\n",
      "sha1": "ed81e4beee8e4f6fa07bebb90f7e3323315b3735"
    },
    {
      "filePath": "kv.thrift",
      "rawContent": "exception ResourceDoesNotExist {\n    1: required string key\n    2: optional string message\n}\n\nservice KeyValue {\n    string getValue(1: string key)\n        throws (1: ResourceDoesNotExist doesNotExist)\n    void setValue(1: string key, 2: string value)\n}\n",
      "sha1": "9e8c1c30d0b6bd7d83426a92962269aaef706295"
    }
  ]
}
```
```sh
yab -p http://localhost:24042 -e json keyvalue yarpc::idls '{ "entryPoint": "many/td/structs.thrift" }' | jq '.body'
{
  "idls": [
    {
      "filePath": "many/td/structs.thrift",
      "includes": [
        "many/td/foo/bar/enums.thrift"
      ],
      "rawContent": "include \"./foo/bar/enums.thrift\"\n\nstruct EmptyStruct {}\n\n//////////////////////////////////////////////////////////////////////////////\n// Structs with primitives\n\nstruct PrimitiveRequiredStruct {\n    1: required bool boolField\n    2: required byte byteField\n    3: required i16 int16Field\n    4: required i32 int32Field\n    5: required i64 int64Field\n    6: required double doubleField\n    7: required string stringField\n    8: required binary binaryField\n}\n\nstruct PrimitiveOptionalStruct {\n    1: optional bool boolField\n    2: optional byte byteField\n    3: optional i16 int16Field\n    4: optional i32 int32Field\n    5: optional i64 int64Field\n    6: optional double doubleField\n    7: optional string stringField\n    8: optional binary binaryField\n}\n\n//////////////////////////////////////////////////////////////////////////////\n// Nested structs (Required)\n\nstruct Point {\n    1: required double x\n    2: required double y\n}\n\nstruct Size {\n    1: required double width\n    2: required double height\n}\n\nstruct Frame {\n    1: required Point topLeft\n    2: required Size size\n}\n\nstruct Edge {\n    1: required Point startPoint\n    2: required Point endPoint\n}\n\nstruct Graph {\n    1: required list<Edge> edges\n}\n\n//////////////////////////////////////////////////////////////////////////////\n// Nested structs (Optional)\n\nstruct ContactInfo {\n    1: required string emailAddress\n}\n\nstruct User {\n    1: required string name\n    2: optional ContactInfo contact\n}\n\n//////////////////////////////////////////////////////////////////////////////\n// self-referential struct\n\ntypedef Node List\n\nstruct Node {\n    1: required i32 value\n    2: optional List tail\n}\n\n\n//////////////////////////////////////////////////////////////////////////////\n// Default values\n\nstruct DefaultsStruct {\n    1: required i32 requiredPrimitive = 100\n    2: optional i32 optionalPrimitive = 200\n\n    3: required enums.EnumDefault requiredEnum = enums.EnumDefault.Bar\n    4: optional enums.EnumDefault optionalEnum = 2\n\n    5: required list<string> requiredList = [\"hello\", \"world\"]\n    6: optional list<double> optionalList = [1, 2.0, 3]\n\n    7: required Frame requiredStruct = {\n        \"topLeft\": {\"x\": 1, \"y\": 2},\n        \"size\": {\"width\": 100, \"height\": 200},\n    }\n    8: optional Edge optionalStruct = {\n        \"startPoint\": {\"x\": 1, \"y\": 2},\n        \"endPoint\":   {\"x\": 3, \"y\": 4},\n    }\n}\n",
      "sha1": "ed81e4beee8e4f6fa07bebb90f7e3323315b3735"
    },
    {
      "filePath": "many/td/foo/bar/enums.thrift",
      "rawContent": "enum EmptyEnum {}\n\nenum EnumDefault {\n    Foo, Bar, Baz\n}\n\nenum EnumWithValues {\n    X = 123,\n    Y = 456,\n    Z = 789,\n}\n\nenum EnumWithDuplicateValues {\n    P, // 0\n    Q = -1,\n    R, // 0\n}\n\n// enum with item names conflicting with those of another enum\nenum EnumWithDuplicateName {\n    A, B, C, P, Q, R, X, Y, Z\n}\n\n// Enum treated as optional inside a struct\nstruct StructWithOptionalEnum {\n    1: optional EnumDefault e\n}\n\nenum RecordType {\n  NAME,\n  HOME_ADDRESS,\n  WORK_ADDRESS\n}\n\nenum lowerCaseEnum {\n    containing, lower_case, items\n}\n",
      "sha1": "cfb1ae035ecf1225722ed73500dbaad7e30bb557"
    }
  ]
}
```
```sh
yab -p http://localhost:24042 -e json keyvalue yarpc::idls '{ "selection": ["many/td/structs.thrift"] }' | jq '.body'
{
  "idls": [
    {
      "filePath": "many/td/structs.thrift",
      "includes": [
        "many/td/foo/bar/enums.thrift"
      ],
      "rawContent": "include \"./foo/bar/enums.thrift\"\n\nstruct EmptyStruct {}\n\n//////////////////////////////////////////////////////////////////////////////\n// Structs with primitives\n\nstruct PrimitiveRequiredStruct {\n    1: required bool boolField\n    2: required byte byteField\n    3: required i16 int16Field\n    4: required i32 int32Field\n    5: required i64 int64Field\n    6: required double doubleField\n    7: required string stringField\n    8: required binary binaryField\n}\n\nstruct PrimitiveOptionalStruct {\n    1: optional bool boolField\n    2: optional byte byteField\n    3: optional i16 int16Field\n    4: optional i32 int32Field\n    5: optional i64 int64Field\n    6: optional double doubleField\n    7: optional string stringField\n    8: optional binary binaryField\n}\n\n//////////////////////////////////////////////////////////////////////////////\n// Nested structs (Required)\n\nstruct Point {\n    1: required double x\n    2: required double y\n}\n\nstruct Size {\n    1: required double width\n    2: required double height\n}\n\nstruct Frame {\n    1: required Point topLeft\n    2: required Size size\n}\n\nstruct Edge {\n    1: required Point startPoint\n    2: required Point endPoint\n}\n\nstruct Graph {\n    1: required list<Edge> edges\n}\n\n//////////////////////////////////////////////////////////////////////////////\n// Nested structs (Optional)\n\nstruct ContactInfo {\n    1: required string emailAddress\n}\n\nstruct User {\n    1: required string name\n    2: optional ContactInfo contact\n}\n\n//////////////////////////////////////////////////////////////////////////////\n// self-referential struct\n\ntypedef Node List\n\nstruct Node {\n    1: required i32 value\n    2: optional List tail\n}\n\n\n//////////////////////////////////////////////////////////////////////////////\n// Default values\n\nstruct DefaultsStruct {\n    1: required i32 requiredPrimitive = 100\n    2: optional i32 optionalPrimitive = 200\n\n    3: required enums.EnumDefault requiredEnum = enums.EnumDefault.Bar\n    4: optional enums.EnumDefault optionalEnum = 2\n\n    5: required list<string> requiredList = [\"hello\", \"world\"]\n    6: optional list<double> optionalList = [1, 2.0, 3]\n\n    7: required Frame requiredStruct = {\n        \"topLeft\": {\"x\": 1, \"y\": 2},\n        \"size\": {\"width\": 100, \"height\": 200},\n    }\n    8: optional Edge optionalStruct = {\n        \"startPoint\": {\"x\": 1, \"y\": 2},\n        \"endPoint\":   {\"x\": 3, \"y\": 4},\n    }\n}\n",
      "sha1": "ed81e4beee8e4f6fa07bebb90f7e3323315b3735"
    }
  ]
}
```

File: x/yarpcmeta/service.go [hunk 1]
```diff
@@ -56,9 +58,88 @@ func (m *service) procs(ctx context.Context, body interface{}) (*procsResponse,
...
 }
 
+type idlsQuery struct {
+	EntryPoint string   `json:"entryPoint,omitempty"`
+	Selection  []string `json:"selection,omitempty"`
+}
+
+type idlFile struct {
+	FilePath   string   `json:"filePath"`
+	SHA1       string   `json:"sha1"`
```
Comment (peter-edge) [inline x/yarpcmeta/service.go]: Same comments as other PR re: SHA1, RawContent. Also note this maps to `Module` in the other PR, naming?

File: x/yarpcmeta/service.go [hunk 2]
```diff
@@ -56,9 +58,88 @@ func (m *service) procs(ctx context.Context, body interface{}) (*procsResponse,
...
+
+	// return the flattened tree by default.
+	if rq == nil || (rq.EntryPoint == "" && len(rq.Selection) == 0) {
+		var r idlsResponse
+		procedures.Flatmap(r.Append)
+		return &r, nil
+	}
+
+	if rq.EntryPoint != "" && len(rq.Selection) > 0 {
+		return nil, errors.New(`ask for either an "entryPoint" or a "selection", but not both`)
```
Comment (peter-edge) [inline x/yarpcmeta/service.go]: make error global var?

File: x/yarpcmeta/service.go [hunk 3]
```diff
@@ -56,9 +58,88 @@ func (m *service) procs(ctx context.Context, body interface{}) (*procsResponse,
...
+			}
+			return false
+		})
+		return &r, nil
+	}
+
+	// find and return every idl matching the selection. No includes are returned.
+	var selection map[string]struct{}
+	if len(rq.Selection) > 0 {
+		selection = make(map[string]struct{})
```
Comment (peter-edge) [inline x/yarpcmeta/service.go]: `selection = make(map[string]struct{}, len(rq.Selection))`

**Comment by peter-edge** [comment]
There's a lot of uncovered code, it's going to mess with our code coverage metrics

**Pull request closed without merging**